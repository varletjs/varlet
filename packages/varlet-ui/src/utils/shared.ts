import { TimeData } from '../countdown/props'

export interface CacheInstance<T> {
  cache: T[]

  has(key: T): boolean

  add(key: T): void

  remove(key: T): void

  clear(): void
}

export const toNumber = (val: number | string | boolean | undefined | null): number => {
  if (val == null) return 0

  if (isString(val)) {
    val = parseFloat(val)
    val = Number.isNaN(val) ? 0 : val
    return val
  }

  if (isBool(val)) return Number(val)

  return val
}

export const isHTMLSupportImage = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:image') || /\.(png|jpg|gif|jpeg|svg)$/.test(val)
}

export const isHTMLSupportVideo = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:video') || /\.(mp4|webm|ogg)$/.test(val)
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isBool = (val: unknown): val is boolean => typeof val === 'boolean'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isPlainObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isObject = (val: unknown): val is Record<string, any> => typeof val === 'object' && val !== null

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isURL = (val: string | undefined | null) => {
  if (!val) {
    return false
  }

  return /^(http)|(\.*\/)/.test(val)
}

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start = 0

  return function loop(this: unknown, ...args) {
    const now = Date.now()
    const elapsed = now - start

    if (!start) {
      start = now
    }

    if (timer) {
      window.clearTimeout(timer)
    }

    if (elapsed >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, mustRunDelay - elapsed)
    }
  }
}

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,

    has(key: T) {
      return this.cache.includes(key)
    },

    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },

    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },

    clear() {
      this.cache.length = 0
    },
  }
}

export const linear = (value: number): number => value

export const cubic = (value: number): number => value ** 3

export const easeInOutCubic = (value: number): number =>
  value < 0.5 ? cubic(value * 2) / 2 : 1 - cubic((1 - value) * 2) / 2

export function parseFormat(format: string, time: TimeData): string {
  const scannedTimes = Object.values(time)
  const scannedFormats = ['DD', 'HH', 'mm', 'ss']
  const padValues = [24, 60, 60, 1000]

  scannedFormats.forEach((scannedFormat, index) => {
    if (!format.includes(scannedFormat)) {
      scannedTimes[index + 1] += scannedTimes[index] * padValues[index]
    } else {
      format = format.replace(scannedFormat, String(scannedTimes[index]).padStart(2, '0'))
    }
  })

  if (format.includes('S')) {
    const ms = String(scannedTimes[scannedTimes.length - 1]).padStart(3, '0')

    if (format.includes('SSS')) {
      format = format.replace('SSS', ms)
    } else if (format.includes('SS')) {
      format = format.replace('SS', ms.slice(0, 2))
    } else {
      format = format.replace('S', ms.slice(0, 1))
    }
  }

  return format
}

export const dt = (value: unknown, defaultText: string | undefined) => (value == null ? defaultText : value)

export const inBrowser = () => typeof window !== 'undefined'

export const uniq = (arr: Array<any>) => [...new Set(arr)]

export function kebabCase(str: string): string {
  const ret = str.replace(/([A-Z])/g, ' $1').trim()
  return ret.split(' ').join('-').toLowerCase()
}
