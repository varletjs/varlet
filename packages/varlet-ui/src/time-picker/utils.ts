import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
