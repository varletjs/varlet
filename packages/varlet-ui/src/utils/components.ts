import {
  createApp,
  h,
  getCurrentInstance,
  inject,
  onUnmounted,
  computed,
  provide,
  reactive,
  isVNode,
  onMounted,
  onBeforeUnmount,
  nextTick,
  ref,
  onActivated,
  onDeactivated,
} from 'vue'
import type { Component, VNode, ComputedRef, ComponentInternalInstance, Ref, ComponentPublicInstance } from 'vue'
import { isArray, removeItem } from '@varlet/shared'

export interface MountInstance {
  instance: ComponentPublicInstance
  unmount: () => void
}

export interface ChildrenCounter {
  collect(instance: ComponentInternalInstance): void

  clear(instance: ComponentInternalInstance): void

  instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
  collect(childProvider: C): void

  clear(childProvider: C): void
}

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export function mount(component: Component): MountInstance {
  const app = createApp(component)
  const host = document.createElement('div')
  document.body.appendChild(host)

  return {
    instance: app.mount(host),
    unmount() {
      app.unmount()
      document.body.removeChild(host)
    },
  }
}

export function mountInstance(
  component: any,
  props: Record<string, any> = {},
  eventListener: Record<string, any> = {}
): {
  unmountInstance: () => void
} {
  const Host = {
    setup() {
      return () =>
        h(component, {
          ...props,
          ...eventListener,
        })
    },
  }

  const { unmount } = mount(Host)
  return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
  const vNodes: VNode[] = []

  const flat = (subTree: any) => {
    if (subTree?.component) {
      flat(subTree?.component.subTree)
      return
    }

    if (Array.isArray(subTree?.children)) {
      subTree.children.forEach((child: any) => {
        if (isVNode(child)) {
          vNodes.push(child)

          flat(child)
        }
      })
    }
  }

  flat(subTree)

  return vNodes
}

export function useAtChildrenCounter(key: symbol) {
  const instances: ComponentInternalInstance[] = reactive([])
  const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  const sortInstances = () => {
    const vNodes: any[] = flatVNodes(parentInstance.subTree)

    instances.sort((a, b) => {
      return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
    })
  }

  const collect = (instance: ComponentInternalInstance) => {
    instances.push(instance)
    sortInstances()
  }

  const clear = (instance: ComponentInternalInstance) => {
    removeItem(instances, instance)
  }

  provide<ChildrenCounter>(key, {
    collect,
    clear,
    instances,
  })

  const length: ComputedRef<number> = computed(() => instances.length)

  return {
    length,
  }
}

export function useAtParentIndex(key: symbol) {
  if (!keyInProvides(key)) {
    return { index: null }
  }

  const childrenCounter: ChildrenCounter = inject<ChildrenCounter>(key) as ChildrenCounter

  const { collect, clear, instances } = childrenCounter

  const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  onMounted(() => {
    nextTick().then(() => collect(instance))
  })
  onUnmounted(() => {
    nextTick().then(() => clear(instance))
  })

  const index = computed(() => instances.indexOf(instance))

  return {
    index,
  }
}

export function useChildren<P, C>(key: symbol) {
  const childProviders: C[] = []

  const collect = (childProvider: C) => {
    childProviders.push(childProvider)
  }

  const clear = (childProvider: C) => {
    removeItem(childProviders, childProvider)
  }

  const bindChildren = (parentProvider: P) => {
    provide<P & BaseParentProvider<C>>(key, {
      collect,
      clear,
      ...parentProvider,
    })
  }

  return {
    childProviders,
    bindChildren,
  }
}

export function useParent<P, C>(key: symbol) {
  if (!keyInProvides(key)) {
    return {
      parentProvider: null,
      bindParent: null,
    }
  }

  const rawParentProvider = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>

  const { collect, clear, ...parentProvider } = rawParentProvider

  const bindParent = (childProvider: C) => {
    onMounted(() => collect(childProvider))
    onBeforeUnmount(() => clear(childProvider))
  }

  return {
    parentProvider,
    bindParent,
  }
}

export function keyInProvides(key: symbol) {
  const instance = getCurrentInstance() as any

  return key in instance.provides
}

export function useValidation() {
  const errorMessage: Ref<string> = ref('')

  const validate = async (rules: any, value: any, apis?: any): Promise<boolean> => {
    if (!isArray(rules) || !rules.length) {
      return true
    }

    const resArr = await Promise.all(rules.map((rule) => rule(value, apis)))

    return !resArr.some((res) => {
      if (res !== true) {
        errorMessage.value = String(res)
        return true
      }

      return false
    })
  }

  const resetValidation = () => {
    errorMessage.value = ''
  }

  const validateWithTrigger = async <T>(validateTrigger: T[], trigger: T, rules: any, value: any, apis?: any) => {
    if (validateTrigger.includes(trigger)) {
      ;(await validate(rules, value, apis)) && (errorMessage.value = '')
    }
  }

  return {
    errorMessage,
    validate,
    resetValidation,
    validateWithTrigger,
  }
}

export function useRouteListener(cb: () => void) {
  onMounted(() => {
    window.addEventListener('hashchange', cb)
    window.addEventListener('popstate', cb)
  })
  onUnmounted(() => {
    window.removeEventListener('hashchange', cb)
    window.removeEventListener('popstate', cb)
  })
}

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)

  onActivated(() => {
    disabled.value = false
  })

  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled,
  }
}

export function exposeApis<T = Record<string, any>>(apis: T) {
  const instance = getCurrentInstance()
  if (instance) {
    Object.assign(instance.proxy!, apis)
  }
}

type ClassName = string | undefined | null
type Classes = (ClassName | [any, ClassName, ClassName?])[]

export function createNamespace(name: string) {
  const namespace = `var-${name}`

  const createBEM = (suffix?: string): string => {
    if (!suffix) return namespace

    return suffix.startsWith('--') ? `${namespace}${suffix}` : `${namespace}__${suffix}`
  }

  const classes = (...classes: Classes): any[] => {
    return classes.map((className) => {
      if (isArray(className)) {
        const [condition, truthy, falsy = null] = className
        return condition ? truthy : falsy
      }

      return className
    })
  }

  return {
    n: createBEM,
    classes,
  }
}

export function call<F extends (...arg: any) => any, P extends Parameters<F>>(
  fn?: F | null,
  ...arg: P
): ReturnType<F> | undefined {
  if (fn) return fn(...arg)
}
