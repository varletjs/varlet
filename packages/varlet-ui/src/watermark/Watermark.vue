<template>
  <div :class="n()">
    <slot />
    <div
      ref="containerRef"
      :class="classes(n('container'), [fullscreen, n('--fullscreen')])"
      :style="{
        backgroundImage: `url(${watermarkUrl})`,
        zIndex: zIndex,
      }"
    >
      <div style="display: none" ref="svgRef">
        <svg
          :viewBox="`0 0 ${width + gapX} ${height + gapY}`"
          :width="`${width + gapX}`"
          :height="`${height + gapY}`"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          :style="{
            padding: `0 ${gapX}px ${gapY}px 0`,
            opacity: opacity,
          }"
        >
          <foreignObject v-if="showContent()" x="0" y="0" :width="width" :height="height">
            <div
              xmlns="http://www.w3.org/1999/xhtml"
              :style="{
                transform: `translate(${offsetX}px, ${offsetY}px) rotate(${rotate}deg)`,
                transformOrigin: 'center',
              }"
            >
              <slot name="content">
                <span :style="{ ...font, fontSize: `${font.fontSize}px`, color: textColor }">{{ content }}</span>
              </slot>
            </div>
          </foreignObject>
          <image
            v-if="!$slots.content && image"
            :href="imageUrl"
            :xlink:href="imageUrl"
            :x="offsetX"
            :y="offsetY"
            :width="width"
            :height="height"
            :style="{
              transformOrigin: 'center',
              transform: `rotate(${rotate}deg)`,
            }"
          />
        </svg>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, type Ref, watch, nextTick, onUnmounted, onMounted } from 'vue'

import { createNamespace } from '../utils/components'
import { props } from './props'
import { getStyle } from '../utils/elements'

const { n, classes } = createNamespace('watermark')

export default defineComponent({
  name: 'VarWatermark',
  props,
  setup(props, { slots }) {
    const watermarkUrl: Ref<string> = ref('')
    const imageUrl: Ref<string> = ref('')
    const svgRef: Ref<SVGElement | null> = ref(null)
    const containerRef: Ref<Element | null> = ref(null)
    const textColor: Ref<string> = ref('')

    const showContent = () => {
      // show slot content first
      if (slots.content) {
        return true
      }

      // if image isn't set, content will take effect
      if (props.content && !props.image) {
        return true
      }

      return false
    }

    const imageToBase64 = async (): Promise<string> =>
      new Promise((resolve) => {
        const canvas = document.createElement('canvas')
        const ctx = canvas.getContext('2d')

        const img = new Image()
        img.crossOrigin = 'anonymous'
        img.referrerPolicy = 'no-referrer'
        img.src = props.image

        img.onload = () => {
          canvas.width = img.width
          canvas.height = img.height
          ctx!.drawImage(img, 0, 0)
          resolve(canvas.toDataURL())
        }
      })

    const svgToBlobUrl = (svgStr: string) => {
      const svgBlob = new Blob([svgStr], {
        type: 'image/svg+xml',
      })

      return URL.createObjectURL(svgBlob)
    }

    const revokeWatermarkUrl = () => {
      if (watermarkUrl.value) {
        URL.revokeObjectURL(watermarkUrl.value)
      }
    }

    // expose
    const resize = async () => {
      textColor.value = getStyle(containerRef.value as Element).color

      if (props.image) {
        imageUrl.value = await imageToBase64()
      }

      await nextTick()

      revokeWatermarkUrl()
      watermarkUrl.value = svgToBlobUrl(svgRef.value!.innerHTML)
    }

    watch(
      () => [
        props.image,
        props.font,
        props.content,
        props.height,
        props.width,
        props.rotate,
        props.gapX,
        props.gapY,
        props.offsetX,
        props.offsetY,
        props.opacity,
      ],
      resize,
      {
        deep: true,
      }
    )

    onMounted(resize)

    onUnmounted(revokeWatermarkUrl)

    return {
      n,
      classes,
      svgRef,
      containerRef,
      watermarkUrl,
      imageUrl,
      textColor,
      showContent,
      resize,
    }
  },
})
</script>

<style lang="less">
@import './watermark';
</style>
