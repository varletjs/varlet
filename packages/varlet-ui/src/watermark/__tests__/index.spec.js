import Watermark from '..'
import VarWatermark from '../Watermark'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'
import { mockCreateElement } from '../../utils/jest'

// simulate URL.createObjectURL() method
global.URL.createObjectURL = jest.fn()

mockCreateElement()

test('test watermark plugin', () => {
  const app = createApp({}).use(Watermark)
  expect(app.component(Watermark.name)).toBeTruthy()
})

test('test watermark content', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      content: 'Varlet',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg span').text()).toBe('Varlet')
  wrapper.unmount()
})

test('test watermark font', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      content: 'Varlet',
      font: {
        fontSize: 15,
        fontWeight: 700,
      },
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark z-index', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      zIndex: 100,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('.var-watermark__container').attributes('style')).toContain('z-index: 100')
  wrapper.unmount()
})

test('test watermark fullscreen', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      fullscreen: true,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('.var-watermark__full')).toBeTruthy()
  wrapper.unmount()
})

test('test watermark opacity', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      opacity: 1,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg').attributes('style')).toContain('opacity: 1')
  wrapper.unmount()
})

test('test watermark image', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark content rotate', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      rotate: 20,
      content: 'Varlet',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg div').attributes('style').includes('rotate(20deg)')).toBeTruthy()
  await wrapper.setProps({
    rotate: 30,
  })
  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg div').attributes('style').includes('rotate(30deg)')).toBeTruthy()
  wrapper.unmount()
})

test('test watermark image rotate', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      rotate: 20,
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg image').attributes('style').includes('rotate(20deg)')).toBeTruthy()
  await wrapper.setProps({
    rotate: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg image').attributes('style').includes('rotate(30deg)')).toBeTruthy()
  wrapper.unmount()
})

test('test watermark width', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      width: 150,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.setProps({
    width: 200,
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark height', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      height: 150,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.setProps({
    height: 200,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark gapX', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      gapX: 20,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg').attributes('style')).toContain('padding: 0px 20px 0px 0px')
  await wrapper.setProps({
    gapX: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg').attributes('style')).toContain('padding: 0px 30px 0px 0px')
  wrapper.unmount()
})

test('test watermark gapY', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      gapY: 20,
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg').attributes('style')).toContain('padding: 0px 0px 20px 0px')
  await wrapper.setProps({
    gapY: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg').attributes('style')).toContain('padding: 0px 0px 30px 0px')
  wrapper.unmount()
})

test('test watermark content offsetX', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      offsetX: 20,
      content: 'Varlet',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg div').attributes('style')).toContain('transform: translate(20px, 0px)')
  await wrapper.setProps({
    offsetX: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg div').attributes('style')).toContain('transform: translate(30px, 0px)')
  wrapper.unmount()
})

test('test watermark content offsetY', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      offsetY: 20,
      content: 'Varlet',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg div').attributes('style')).toContain('transform: translate(0px, 20px)')
  await wrapper.setProps({
    offsetY: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.find('svg div').attributes('style')).toContain('transform: translate(0px, 30px)')
  wrapper.unmount()
})

test('test watermark image offsetX', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      offsetX: 20,
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.setProps({
    offsetX: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark image offsetY', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      offsetY: 20,
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.setProps({
    offsetY: 30,
  })

  await wrapper.vm.$nextTick()
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test watermark content slot', async () => {
  const wrapper = mount(VarWatermark, {
    slots: {
      content: () => 'Varlet UI',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg div').text()).toBe('Varlet UI')
  wrapper.unmount()
})

test('test watermark priority between image and content', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
      content: 'Varlet UI',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg image').exists()).toBeTruthy()
  expect(wrapper.find('svg span').exists()).toBe(false)
  wrapper.unmount()
})

test('test watermark priority between image and content slot', async () => {
  const wrapper = mount(VarWatermark, {
    props: {
      image: 'https://varlet.gitee.io/varlet-ui/varlet_icon.png',
    },
    slots: {
      content: () => 'Varlet UI',
    },
  })

  await wrapper.vm.$nextTick()

  expect(wrapper.find('svg image').exists()).toBe(false)
  expect(wrapper.find('svg div').text()).toBe('Varlet UI')
  wrapper.unmount()
})
