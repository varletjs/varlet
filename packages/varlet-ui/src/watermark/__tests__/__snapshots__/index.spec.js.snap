// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test watermark font 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <foreignObject x="0" y="0" width="100" height="100">
          <div xmlns="http://www.w3.org/1999/xhtml" style="transform: translate(0px, 0px) rotate(-22deg); transform-origin: center;"><span style="font-size: 15px; font-weight: 700;">Varlet</span></div>
        </foreignObject>
        <!--v-if-->
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark height 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 150" width="100" height="150" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <!--v-if-->
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark height 2`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(undefined); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 200" width="100" height="200" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <!--v-if-->
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark image 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <image href="" xlink:href="" x="0" y="0" width="100" height="100" style="transform: rotate(-22deg); transform-origin: center;"></image>
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark image offsetX 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <image href="" xlink:href="" x="20" y="0" width="100" height="100" style="transform: rotate(-22deg); transform-origin: center;"></image>
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark image offsetX 2`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <image href="" xlink:href="" x="30" y="0" width="100" height="100" style="transform: rotate(-22deg); transform-origin: center;"></image>
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark image offsetY 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <image href="" xlink:href="" x="0" y="20" width="100" height="100" style="transform: rotate(-22deg); transform-origin: center;"></image>
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark image offsetY 2`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" width="100" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <image href="" xlink:href="" x="0" y="30" width="100" height="100" style="transform: rotate(-22deg); transform-origin: center;"></image>
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark width 1`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 150 100" width="150" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <!--v-if-->
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;

exports[`test watermark width 2`] = `
"<div class="var-watermark">
  <!--teleport start-->
  <div class="var-watermark__container" style="background-image: url(undefined); z-index: 10;">
    <div style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 200 100" width="200" height="100" style="padding: 0px 0px 0px 0px; opacity: 0.5;">
        <!--v-if-->
        <!--v-if-->
      </svg></div>
  </div>
  <!--teleport end-->
</div>"
`;
